Add a way to batch spawn tasks

notgull opened the pull request:
For some workloads many tasks are spawned at a time. This requires
locking and unlocking the executor's inner lock every time you spawn a
task. If you spawn many tasks this can be expensive.

This commit exposes a new "spawn_many" method on both types. This
method allows the user to spawn an entire set of tasks at a time.

Closes #91

james7132 commented on src/lib.rs:
```diff
@@ -417,6 +493,76 @@ impl<'a> LocalExecutor<'a> {
+        let mut active = self.inner().state().active.lock().unwrap();
+
+        for future in futures {
+            // Remove the task from the set of active tasks when the future finishes.
```
The first part of this loop looks identical to the corresponding part in `spawn`, perhaps refactor this out to a common private function?

james7132 commented on src/lib.rs:
```diff
@@ -417,6 +493,76 @@ impl<'a> LocalExecutor<'a> {
+    /// [`Executor::spawn_many`]: Executor::spawn_many
+    pub fn spawn_many<T: Send + 'a, F: Future<Output = T> + Send + 'a>(
+        &self,
+        futures: impl IntoIterator<Item = F>,
```
This forces the futures to be the same type and also requires something to collect the tasks. If we're ok with users potentially injecting extra non-blocking code via iterator implementations (with the doc comments as a deterrent), wouldn't a `SpawnScope<'a>` be a more flexible API?

notgull replied:
It's much harder to accidentally create a blocking `Iterator` than it is to accidentally hold a lock type across an `await` point. I've seen the former a couple of times and the latter all the time.

james7132 commented on src/lib.rs:
````diff
@@ -170,6 +170,82 @@ impl<'a> Executor<'a> {
+    /// ```
+    ///
+    /// [`spawn`]: Executor::spawn
+    pub fn spawn_many<T: Send + 'a, F: Future<Output = T> + Send + 'a>(
````
The PR description says it's named `spawn_batch` but it's named differently here. Is that intentional?

notgull replied:
Fixed the PR description

james7132 commented on src/lib.rs:
```diff
@@ -417,6 +493,76 @@ impl<'a> LocalExecutor<'a> {
+        futures: impl IntoIterator<Item = F>,
+        handles: &mut impl Extend<Task<F::Output>>,
+    ) {
+        let mut active = self.inner().state().active.lock().unwrap();
```
For a local executor, why does it need a lock at all? Probably not necessary to address in this PR, but I thought the local executor itself could be `!Send` and `!Sync`, so wouldn't a RefCell be sufficient for a LocalExecutor state?

notgull replied:
It would require a reimplementation of `LocalExecutor`'s internals, which is something I don't feel like doing right now as the current implementation works well enough. I would accept a PR for this.

fogti replied:
ok, as a change this would propagate through the entire code base because it would require a different `State` struct to actually eliminate all the mutexes, etc, which then probably needs either a duplication of all the other internal structs which hold references to `State`, or an abstraction over the two variants of `State`.

james7132 replied:
It's probably doable with `macro_rules`, but that might have a serious impact on readability.

fogti replied:
I think a proper abstraction would require a lot of interior mutability, we just can't directly abstract all that without pulling in niche dependencies, I suppose..

notgull replied:
Generally thread unsafe primitives are provided by `unsend` instead of `smol`. In this case see [`unsend::Executor`](https://docs.rs/unsend/latest/unsend/executor/struct.Executor.html)

james7132 commented on src/lib.rs:
```diff
@@ -170,6 +170,82 @@ impl<'a> Executor<'a> {
+            let (runnable, task) = unsafe {
+                Builder::new()
+                    .propagate_panic(true)
+                    .spawn_unchecked(|()| future, self.schedule())
```
Probably out of the scope of this PR, but would there be a reasonable way to add all of the tasks to the injector queue in one go instead of one at a time?

fogti replied:
it might be good to keep in mind what `spawn_unchecked` does: https://docs.rs/async-task/latest/src/async_task/runnable.rs.html#504-533
Here, it would mean inlining `self.schedule()` basically, which does:
```rust
        move |runnable| {
            state.queue.push(runnable).unwrap();
            state.notify();
        }
```
and we can't really do batch processing in `concurrent-queue` yet, so it would just mean that `state.notify()` gets called once instead of: for every task.

fogti replied:
Problem is that we still need to do the `notify` anyways in the supplied `schedule` function so that later wake-ups work properly, so doing this properly means first patching in a way for batch processing into `async-task` afaik

notgull replied:
Yes, right now `concurrent-queue` does not support batch inserts. Making it support batch inserts would require a rewrite of the algorithm, which I would not support.

fogti replied:
ok, so we can deem this "out of scope because it would require a large-scale rewrite across multiple projects"

james7132 commented on src/lib.rs:
```diff
@@ -942,6 +1054,17 @@ fn debug_executor(executor: &Executor<'_>, name: &str, f: &mut fmt::Formatter<'_
+/// Container with one item.
+///
+/// This implements `Extend` for one-off cases.
+struct Container<T>(Option<T>);
```
This doesn't seem to be used.

james7132 commented on src/lib.rs:
```diff
@@ -397,24 +470,63 @@ impl<'a> LocalExecutor<'a> {
-        task
+        for future in futures {
+            // SAFETY: F and T are both `Send`.
+            handles.extend(Some(unsafe {
```
`std::iter::once` may be preferable to `Option`, though I assume the branch it's iterator implementation gets optimized out at some point.

notgull replied:
I've realized that we can just convert `futures` into an `Iterator` of `Task`s, and then use `handles.extend` with that iterator. That should probably enable more optimizations.

notgull commented:
@smol-rs/admins I would appreciate a review on this, as it's blocking performance work in Bevy.

taiki-e commented:
A common bug with *_many functions in Rust is that memory safety is violated when the iterator panics. So could you please add a Miri-compatible test for such a case? (I do not believe that this implementation violates memory safety on that case, but at least the implementation does not seem to care about cases where the iterator panics.)
Otherwise it looks ok to me.

notgull commented:
> A common bug with *_many functions in Rust is that memory safety is violated when the iterator panics. So could you please add a Miri-compatible test for such a case? (I do not believe that this implementation violates memory safety on that case, but at least the implementation does not seem to care about cases where the iterator panics.) Otherwise it looks ok to me.

Added! I've also added a workaround for the mutex poisoning that shows up here.

taiki-e commented:
It seems miri reported memory leaks.
```
error: memory leaked: alloc44788 (Rust heap, size: 120, align: 8), allocated here:
```
https://github.com/smol-rs/async-executor/actions/runs/7999577501/job/21847610801?pr=92

notgull commented:
Hmm, looks like the issue was related to the thread-optimized backend? It looks like it doesn't leak memory now.
